R/chat.R

```diff
@@ -412,22 +447,34 @@ ChatOpenAI <- R6::R6Class("ChatOpenAI",
 #' @export
 print.ChatOpenAI <- function(x, ...) {
   cat("<ChatOpenAI>\n")
-  messages <- x$.__enclos_env__$private$messages
-  for (message in messages) {
+  for (message in x$messages()) {
     color <- switch(message$role,
-      user = "blue",
-      system = ,
-      assistant = "green"
+      user = cli::col_blue,
+      assistant = cli::col_green,
+      identity
     )
-    cat(cli::rule(message$role, col = color), "\n", sep = "")
-    cat(message$content, "\n", sep = "")
+    cli::cli_rule("{color(message$role)}")
+    if (!is.null(message$content)) {
+      # Using cli_text for word wrapping. Passing `"{message$content}"` instead of
+      # `message$content` to avoid evaluation of the (potentially malicious)
+      # content.
+      cli::cli_text("{message$content}")
+    }
+    if (!is.null(message$tool_calls)) {
+      cli::cli_text("Tool calls:")
+      for (tool_call in message$tool_calls) {
```

Comment:
Just convert back to JSON and print it?

Reply:
Yeah, that was my first instinct...

1. The `tool_calls` can't be printed because it contains random IDs--well, we could, but it makes snapshot testing annoying.
2. We could print each of the tool calls' `function` which would result in `{"name": "rnorm", "arguments": "{\"n\": 5}}` which bothered me because of the noisy double quotes in the key names, but especially, the double-JSON-encoding for `arguments` (because it's actually a string that just happens to be valid JSON).

This way, we get `rnorm({"n": 5})` (which even that, I was tempted to turn into `rnorm(n = 5)`.

Reply:
Ah makes sense. For your last point, you could do `call2(funcname, !!!args)` to make a call object which should (I think) print nicely)